osgp/platform/osgp-adapter-domain-core/src/main/java/org/opensmartgridplatform/adapter/domain/core/application/services/FirmwareManagementService.java

```diff
@@ -89,13 +107,74 @@ public void updateFirmware(final CorrelationIds ids,
         this.findOrganisation(ids.getOrganisationIdentification());
         final Device device = this.findActiveDevice(ids.getDeviceIdentification());
 
+        if (device instanceof Ssld) {
+            final Ssld ssld = this.findSsldForDevice(device);
+            this.createSsldPendingFirmwareUpdateRecord(ids, ssld,
+                    firmwareUpdateMessageDataContainer.getFirmwareModuleData(),
+                    firmwareUpdateMessageDataContainer.getFirmwareUrl());
+        }
+
         this.osgpCoreRequestMessageSender.send(
                 new RequestMessage(ids,
                         this.domainCoreMapper.map(firmwareUpdateMessageDataContainer,
                                 org.opensmartgridplatform.dto.valueobjects.FirmwareUpdateMessageDataContainer.class)),
                 messageType, messagePriority, device.getIpAddress(), scheduleTime);
     }
 
+    private void createSsldPendingFirmwareUpdateRecord(final CorrelationIds ids, final Ssld ssld,
+            final FirmwareModuleData firmwareModuleData, final String firmwareUrl) {
+        try {
+            final String[] split = firmwareUrl.split("/");
+            Assert.isTrue(split.length >= 1, "Splitting URL on / failed!");
+            final String firmwareFilename = split[split.length - 1];
+
+            final List<FirmwareFile> firmwareFiles = this.firmwareFileRepository.findByFilename(firmwareFilename);
+            Assert.isTrue(firmwareFiles.size() == 1, "Expected 1 firmware file for filename: " + firmwareFilename);
+            final FirmwareFile firmwareFile = firmwareFiles.get(0);
+
+            final FirmwareFileFirmwareModule firmwareFileFirmwareModule = this.firmwareFileFirmwareModuleRepository
+                    .findByFirmwareFile(firmwareFile);
+            final String firmwareVersion = firmwareFileFirmwareModule.getModuleVersion();
+
+            final FirmwareModuleType firmwareModuleType = firmwareModuleData.getFirmwareModuleType();
+
+            SsldPendingFirmwareUpdate ssldPendingFirmwareUpdate = new SsldPendingFirmwareUpdate(true,
+                    firmwareModuleType, firmwareVersion, "CORE", "1.0", ids.getOrganisationIdentification());
+            ssldPendingFirmwareUpdate = this.ssldPendingFirmwareUpdateRepository.save(ssldPendingFirmwareUpdate);
+
+            ssld.setSsldPendingFirmwareUpdate(ssldPendingFirmwareUpdate);
+            this.ssldRepository.save(ssld);
```

Comment:
Thinking about domain driven design, should the SsldPendingFirmwareUpdate class be an entity or value object? Should it be an aggregate root or be part of the SSLD aggregate root?
It feels a bit awkward having to save both ssldPendingFirmwareUpdate and the ssld objects.
We could make it a value object (and save it as an embeddable of an Ssld) or a aggregate root on it's own. (In which case we could just add the device identification as a field in the SsldPendingFirmwareUpdate).